js-pkg/client/src/provider.ts

```diff
@@ -143,27 +151,101 @@ export class YSweetProvider {
     this.awareness.on('update', this.handleAwarenessUpdate.bind(this))
     this.WebSocketPolyfill = extraOptions.WebSocketPolyfill || WebSocket
 
+    this.online = this.online.bind(this)
+    this.offline = this.offline.bind(this)
+    if (typeof window !== 'undefined') {
+      window.addEventListener('offline', this.offline)
+      window.addEventListener('online', this.online)
+    }
+
     doc.on('update', this.update.bind(this))
 
     if (extraOptions.connect !== false) {
       this.connect()
     }
   }
 
+  private offline() {
+    // When the browser indicates that we are offline, we immediately
+    // probe the connection status.
+    // This accelerates the process of discovering we are offline, but
+    // doesn't mean we entirely trust the browser, since it can be wrong
+    // (e.g. in the case that the connection is over localhost).
+    this.checkSync()
+  }
+
+  private online() {
+    if (this.reconnectSleeper) {
+      this.reconnectSleeper.wake()
+    }
+  }
+
+  private clearHeartbeat() {
+    if (this.heartbeatHandle) {
+      clearTimeout(this.heartbeatHandle)
+      this.heartbeatHandle = null
+    }
+  }
+
+  private resetHeartbeat() {
+    this.clearHeartbeat()
+    this.heartbeatHandle = setTimeout(() => {
+      this.checkSync()
+      this.setConnectionTimeout()
```

Comment:
This should no longer be necessary since we now set the timeout in `this.checkSync`
```suggestion
```